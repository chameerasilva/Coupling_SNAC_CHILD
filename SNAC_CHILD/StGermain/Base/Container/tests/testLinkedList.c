
/*~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
**
** Copyright (C), 2003, Victorian Partnership for Advanced Computing (VPAC) Ltd, 110 Victoria Street, Melbourne, 3053, Australia.
**
** Authors:
**	Stevan M. Quenette, Senior Software Engineer, VPAC. (steve@vpac.org)
**	Patrick D. Sunter, Software Engineer, VPAC. (pds@vpac.org)
**	Luke J. Hodkinson, Computational Engineer, VPAC. (lhodkins@vpac.org)
**	Siew-Ching Tan, Software Engineer, VPAC. (siew@vpac.org)
**	Alan H. Lo, Computational Engineer, VPAC. (alan@vpac.org)
**	Raquibul Hassan, Computational Engineer, VPAC. (raq@vpac.org)
**
**  This library is free software; you can redistribute it and/or
**  modify it under the terms of the GNU Lesser General Public
**  License as published by the Free Software Foundation; either
**  version 2.1 of the License, or (at your option) any later version.
**
**  This library is distributed in the hope that it will be useful,
**  but WITHOUT ANY WARRANTY; without even the implied warranty of
**  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
**  Lesser General Public License for more details.
**
**  You should have received a copy of the GNU Lesser General Public
**  License along with this library; if not, write to the Free Software
**  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
**
** $Id: testLinkedList.c 2136 2005-02-28 02:47:13Z RaquibulHassan $
**
**~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/

#include <mpi.h>
#include "Base/Foundation/Foundation.h"
#include "Base/IO/IO.h"
#include "Base/Container/Container.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

#define NUM_DATA 100

int compareFunction(void *data1, void *data2)
{
	int *d1 = NULL, *d2 = NULL;

	d1 = (int*)data1;
	d2 = (int*)data2;

	if (d1 == NULL || d2 == NULL){
		return 0;	
	}
	
	if (*d1 > *d2){
		return  1;
	}
	else if (*d1 == *d2){
		return 0;
	}
	else{
		return -1;
	}
}

void dataPrintFunction( void *nodeData, void *args )
{
	Stream *myStream = NULL;
	
	assert (nodeData);
	myStream = Journal_Register( InfoStream_Type, "LinkedListStream" );

	Journal_Printf( myStream, "\t%d\n", *(int*)nodeData );
}

void dataCopyFunction( void **nodeData, void *newData, SizeT dataSize)
{
	*nodeData = Memory_Alloc_Bytes_Unnamed(dataSize, char);
	memset(*nodeData, 0, dataSize);

	memcpy(*nodeData, newData, dataSize);
}
	
int main( int argc, char* argv[] ) {
	MPI_Comm			CommWorld;
	int				rank;
	int				numProcessors;
	int				procToWatch;
	
	LinkedList *numList = NULL;
	int i = 0;
	int *array[NUM_DATA] = {0};
	int *data;
	int *result = NULL;
	float randomNum, randomMax;
	Stream *myStream = NULL;
	
	/* Initialise MPI, get world info */
	MPI_Init( &argc, &argv );
	MPI_Comm_dup( MPI_COMM_WORLD, &CommWorld );
	MPI_Comm_size( CommWorld, &numProcessors );
	MPI_Comm_rank( CommWorld, &rank );

	BaseFoundation_Init( &argc, &argv );
	BaseIO_Init( &argc, &argv );
	BaseContainer_Init( &argc, &argv );

	if( argc >= 2 ) {
		procToWatch = atoi( argv[1] );
	}
	else {
		procToWatch = 0;
	}
	
	if( rank == procToWatch ) {
		
		myStream = Journal_Register( InfoStream_Type, "LinkedListStream" );
		numList = LinkedList_New(
					compareFunction,
					dataCopyFunction,
					dataPrintFunction,
					NULL,
					LINKEDLIST_UNSORTED);

		Journal_Printf( myStream, "Inserting data into the List\n");
		
		for(i=0; i<NUM_DATA; i++){
			array[i] = Memory_Alloc(int, "TestLinkedList_ArrayEntry");
			randomNum = rand();
			randomMax = RAND_MAX;
			*array[i] = NUM_DATA - i;
			LinkedList_InsertNode(numList, (void*)array[i], sizeof(int));
		}
	
		Journal_Printf(myStream, "\nPrinting the list\n");
		Print( (void*)numList, myStream );
		
		Journal_Printf(myStream, "\nDeleting half the nodes previously inserted into the list\n");
		for(i=0; i<NUM_DATA/2; i++){
			LinkedList_DeleteNode(numList, (void*)array[i]);
		}
		
		Journal_Printf(myStream, "\nPrinting the list\n");
		Print( (void*)numList, myStream );
		

		Journal_Printf(myStream, "\nRetrieving the data held in the list as an array\n");
		data = LinkedList_ReturnArray(numList, int);

		Journal_Printf(myStream, "\nPrinting the array just retrieved\n");
		for(i=0; i<numList->nodeCount; i++){
			printf ("arrayData - %d\n", data[i]);
		}
	
		Journal_Printf( myStream, "\nSearching for Node data in the list\n" );
		for(i=0; i<NUM_DATA/4; i++){
			i = data[i];
			if ((result = LinkedList_FindNodeData(numList, &i, int)) != NULL){
				printf ("Looking for - %d, Found..\n", *result);
			}
			else{
				printf ("Failed to find.. %d\n", i);
			}
		}
	}

	Journal_Printf( myStream, "\nDeleting the List\n" );
	Stg_Class_Delete( (void*)numList );
	
	BaseContainer_Finalise();
	BaseIO_Finalise();
	BaseFoundation_Finalise();
	
	/* Close off MPI */
	MPI_Finalize();
	
	return 0; /* success */
}
