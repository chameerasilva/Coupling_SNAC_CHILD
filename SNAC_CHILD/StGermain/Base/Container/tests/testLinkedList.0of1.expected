Inserting data into the List

Printing the list
Stg_Class (ptr): 0x806b498
	sizeOfSelf: 52
	_deleteSelf: Yes
	type: LinkedList
	_delete (func ptr): 0x8058ff5
	_print (func ptr): 0x8058eb0
	_copy (func ptr): (nil)
LinkedList (ptr): (0x806b498)
	NodeCount		 - 100
	LinkedList Order	 - UNSORTED
	LinkedList data	 - 
	1
	2
	3
	4
	5
	6
	7
	8
	9
	10
	11
	12
	13
	14
	15
	16
	17
	18
	19
	20
	21
	22
	23
	24
	25
	26
	27
	28
	29
	30
	31
	32
	33
	34
	35
	36
	37
	38
	39
	40
	41
	42
	43
	44
	45
	46
	47
	48
	49
	50
	51
	52
	53
	54
	55
	56
	57
	58
	59
	60
	61
	62
	63
	64
	65
	66
	67
	68
	69
	70
	71
	72
	73
	74
	75
	76
	77
	78
	79
	80
	81
	82
	83
	84
	85
	86
	87
	88
	89
	90
	91
	92
	93
	94
	95
	96
	97
	98
	99
	100

Deleting half the nodes previously inserted into the list

Printing the list
Stg_Class (ptr): 0x806b498
	sizeOfSelf: 52
	_deleteSelf: Yes
	type: LinkedList
	_delete (func ptr): 0x8058ff5
	_print (func ptr): 0x8058eb0
	_copy (func ptr): (nil)
LinkedList (ptr): (0x806b498)
	NodeCount		 - 50
	LinkedList Order	 - UNSORTED
	LinkedList data	 - 
	1
	2
	3
	4
	5
	6
	7
	8
	9
	10
	11
	12
	13
	14
	15
	16
	17
	18
	19
	20
	21
	22
	23
	24
	25
	26
	27
	28
	29
	30
	31
	32
	33
	34
	35
	36
	37
	38
	39
	40
	41
	42
	43
	44
	45
	46
	47
	48
	49
	50

Retrieving the data held in the list as an array

Printing the array just retrieved
arrayData - 1
arrayData - 2
arrayData - 3
arrayData - 4
arrayData - 5
arrayData - 6
arrayData - 7
arrayData - 8
arrayData - 9
arrayData - 10
arrayData - 11
arrayData - 12
arrayData - 13
arrayData - 14
arrayData - 15
arrayData - 16
arrayData - 17
arrayData - 18
arrayData - 19
arrayData - 20
arrayData - 21
arrayData - 22
arrayData - 23
arrayData - 24
arrayData - 25
arrayData - 26
arrayData - 27
arrayData - 28
arrayData - 29
arrayData - 30
arrayData - 31
arrayData - 32
arrayData - 33
arrayData - 34
arrayData - 35
arrayData - 36
arrayData - 37
arrayData - 38
arrayData - 39
arrayData - 40
arrayData - 41
arrayData - 42
arrayData - 43
arrayData - 44
arrayData - 45
arrayData - 46
arrayData - 47
arrayData - 48
arrayData - 49
arrayData - 50

Searching for Node data in the list
Looking for - 1, Found..
Looking for - 3, Found..
Looking for - 5, Found..
Looking for - 7, Found..
Looking for - 9, Found..
Looking for - 11, Found..
Looking for - 13, Found..
Looking for - 15, Found..
Looking for - 17, Found..
Looking for - 19, Found..
Looking for - 21, Found..
Looking for - 23, Found..
Looking for - 25, Found..

Deleting the List
